### ISSUE TITLE
Jetifier fails with Facebook SDK 7.17.2

### ISSUE THREAD

**Issue by doom-goober** [googlesamples/unity-jar-resolver#360]
### Please fill in the following fields:
Unity editor version: 2019.3.0f6
External Dependency Manager version: 1.2.135.0
Source you installed EDM (from .unitypackage or Unity Package Manager): .unitypackage
Features in External Dependency Manager in use (Android Resolver, iOS Resolver, VersionHandler, etc.): Jetifier
Plugins SDK in use (Firebase, Admob, Facebook, etc.): Firebase, Facebook
Platform you are using the Unity editor on (Mac, Windows, or Linux): Windows
Platform you are targeting (iOS, Android, and/or desktop): Android
Scripting Runtime (Mono, and/or IL2CPP): Mono

### Please describe the issue here:
1) Install Firebase Unity SDK 6.13 (We have Auth, Functions, and Database)
2) Install Facebook SDK for Unity 7.17.2
3) Turn Jetifier on.
3) Force Resolve.
3) Build and Run onto Android Device.
4) The build will immediate exception with a series of duplicate class exceptions between AndroidX and android.support libraries:
java.lang.RuntimeException: Duplicate class android.support.customtabs.ICustomTabsCallback found in modules classes.jar (androidx.browser:browser:1.0.0) and classes.jar (com.android.support:customtabs:28.0.0)

Other Notes: It appears the Unity Dependency Manager is failing to correctly Jetify the Facebook SDK. However, if you export the project and let Android Studio Jetify the project, everything works fine.

### Please answer the following, if applicable:
What's the issue repro rate? (eg 100%, 1/5 etc)
100%

**Comment by Russel1** [comment]
I have the same problem.

**Comment by patm1987** [comment]
Thanks for reaching out, can y'all do me a favor and verify that the issue occurs in in one of the [Unity quickstarts](https://github.com/firebase/quickstart-unity) (say the [Auth](https://github.com/firebase/quickstart-unity/tree/master/auth/testapp) one).

Can you also verify whether you're using the gradle template. If you use the gradle resolution, you'll likely have the file `Assets/Plugins/Android/mainTemplate.gradle` and you'll see it in your build settings (this varies with Unity version):
![image](https://user-images.githubusercontent.com/1111611/79591120-4f787c00-8095-11ea-822e-ab9bcf7c00e0.png)

You'll also need to have patching enabled:
![image](https://user-images.githubusercontent.com/1111611/79591271-7df65700-8095-11ea-9e00-179bab6a0713.png)

In this setup, you'll want to look at the `mainTemplate.gradle` file after resolving resources and look for duplicates here:
```
dependencies {
    implementation fileTree(dir: 'libs', include: ['*.jar'])
// Android Resolver Dependencies Start
    implementation 'com.google.firebase:firebase-analytics:17.2.2' // Assets/Firebase/Editor/AppDependencies.xml:15
    implementation 'com.google.firebase:firebase-analytics-unity:6.13.0' // Assets/Firebase/Editor/AnalyticsDependencies.xml:18
    implementation 'com.google.firebase:firebase-app-unity:6.13.0' // Assets/Firebase/Editor/AppDependencies.xml:20
    implementation 'com.google.firebase:firebase-common:19.3.0' // Assets/Firebase/Editor/AppDependencies.xml:13
// Android Resolver Dependencies End
**DEPS**}
```

and you can prune dependencies here as a workaround. It would be helpful to know if you see the erroneous entries here.

If you're not using this process (you see a bunch of `jar` files in `Assets/Plugins/Android`) let us know as well.

Thanks for your help!
--Patrick

**Comment by chkuang-g** [comment]
@doom-goober 
I think this is an issue in Facebook SDK.
https://github.com/facebook/facebook-sdk-for-unity/issues/303

**Comment by gromilQaaaa** [comment]
https://documentation.onesignal.com/docs/troubleshooting-unity#section-android-x-compatibility

Didn't try yet, but found possible fix on onesignal help section
![image](https://user-images.githubusercontent.com/5392680/79975658-01ec7c80-84a4-11ea-8873-da81ea2b6ade.png)

**Comment by Russel1** [comment]
@gromilQaaaa Yeah thanks a lot, the OneSignal solution fixed the problem. :)

**Comment by doom-goober** [comment]
@chkuang-g @patm1987 

- With all the suggested settings, using "Build and Run" from Unity, Firebase Auth sample works fine.
- With all the suggested settings, using "Build and Run" from Unity, my app with FacebookSDK and Firebase, fails with "duplicate classes".
- From exported project to Android Studio, my app with Facebook and Firebase runs on device.
- With all the suggested settings, using "Build and Run" from unity, my app with FacebookSDK and Firebase runs on device if I use @gromilQaaaa suggestion of adding settingsTemplate.gradle.

This leads me to conclude that problem is the Jetifier is failing with a combo of Facebook and Unity. I opened this bug because I thought that the Jetifier in Unity was owned by the Unity Jar Resolver team. I apologize if I opened the bug in the wrong place. Who owns the Jetifier Process in Unity? Is that a Unity thing or a Google thing? Should I close this bug and file it somewhere else?

Thanks! (Sorry for delay in response. My team is approaching a ship date.)

**Comment by hlongvu** [comment]
I follow OneSignal solution to fix then can build successfully. But the apk size increase significantly. Is this due to androidx library size or something else?
Does anyone face the same issue?

**Comment by Russel1** [comment]
I agree my apk size has also increased by 8.8 MB.

**Comment by doom-goober** [comment]
@chkuang-g @patm1987 Do you still need more info? This appears to be an interplay between Unity, how the resolver runs the Jetifier/AndroidX, and the Facebook SDK.

The workaround everyone has settled on (see the two comments above this one) is to enable the jetifier and androidX using a gradle script rather than through the resolver as per the instructions here: https://documentation.onesignal.com/docs/troubleshooting-unity#section-android-x-compatibility

That's enough info, right? The resolver appears to NOT be interacting with gradle correctly but gradle talking to gradle works correctly.

Please let me know if you need anything else.

**Comment by mohamad-al-amaary** [comment]
Using newest 2019.3 official version (mine is 2019.3.12f1 atm) you can just use Custom Gradle Properties Template on top of the Main Gradle Template, and add:
`android.useAndroidX=true
android.enableJetifier=true`
to the newly generated gradleTemplate.properties file.

I have been exporting to Android studio to build with all the SDKs my publisher wants and working through the miseries of Android build errors ever since, as a junior dev I think I encountered every issue, but now I can finally build from within Unity. Jetifier wasn't working with Facebook SDK properly except in Android studio, so if you used enough AndroidX migrated plugins it would be hard to build from within Unity without some hacky changes.

I also usually have to change Plugins/Android/AndroidManifest.xml to have this in application tag:
`tools:replace="android:appComponentFactory" android:appComponentFactory="leavemealone"`
also `xmlns:tools="http://schemas.android.com/tools"` in manifest tag

Yes, it's a random string as I'm not sure what the purpose of this is but it works and fixes my manifest merger build error, oh well.

I can now build with Ironsource (many adapters with it), Facebook SDK, GameAnalytics, Adjust, Firebase, whatever else. Basically just get latest Unity Jar Resolver, check settings remove auto resolve, resolve on build, set patch maintemplate and use jetifier to true, force resolve, add any extra dependencies/repos you want, do the gradleproperties trick, build within Unity, done.

**Comment by Russel1** [comment]
@mohamad-al-amaary @doom-goober Has your APK size also increased significantly with the latest Unity?

**Comment by mohamad-al-amaary** [comment]
Same size @Russel1

**Comment by ldrly-roger** [comment]
I cannot thank you enough @mohamad-al-amaary, your advice works for me and has unblocked our progress! 👏 👏 👏

**Comment by chkuang-g** [comment]
@mohamad-al-amaary 

Thank you for offering the workaround for this issue.

This is definitely a change of behavior from Unity 2019.3+.  We can investigate how to automate this process in EDM4U, or alternative way to solve this. We also welcome any pull request to solve this issue!

**Comment by mohamad-al-amaary** [comment]
Easiest way to build now as I mentioned in my previous answer with latest 2019.3 is enabling and using gradleTemplate.properties and adding in 
`android.useAndroidX=true
android.enableJetifier=true`
since those properties in mainTemplate just don't work.

So maybe automation would be making the "Use Jetifier" setting of EDM4U enable custom gradle properties in Unity and automatically add those in, but this feature is only available in 2019.3.9+ I think (definitely not sure but something like that) - for previous versions of 2019.3 I guess the OneSignal solution is the only way other than the dreadful export to Android Studio option (it's not so bad but hella time consuming if you wanna make a lot of builds and work within a team with testers and marketing builds etc.)
